src/lib.rs

```diff
@@ -1,101 +1,405 @@
+    let pending_fixes: Mutex<HashMap<RuleName, Vec<PendingFix>>> = Default::default();
+    tree_sitter_grep::run_for_slice_with_callback(
+        file_contents,
+        tree_sitter_grep_args.clone(),
+        |capture_info| {
+            let (rule, rule_listener) =
+                aggregated_queries.get_rule_and_listener(capture_info.pattern_index);
+            let mut query_match_context =
+                QueryMatchContext::new(path, file_contents, rule, &config);
+            (rule_listener.on_query_match)(capture_info.node, &mut query_match_context);
+            if let Some(reported_violations) = query_match_context.violations.take() {
+                violations.lock().unwrap().extend(reported_violations);
+            }
+            if let Some(fixes) = query_match_context.into_pending_fixes() {
+                pending_fixes
+                    .lock()
+                    .unwrap()
+                    .entry(rule.meta.name.clone())
+                    .or_default()
+                    .extend(fixes);
+            }
+        },
+    )
+    .unwrap();
+    let mut violations = violations.into_inner().unwrap();
+    let pending_fixes = pending_fixes.into_inner().unwrap();
+    if pending_fixes.is_empty() {
+        return violations;
+    }
+    run_fixing_loop(
+        &mut violations,
+        file_contents,
+        pending_fixes,
+        tree_sitter_grep_args,
+        &aggregated_queries,
+        path,
+        &config,
+    );
+    violations
+}
+
+fn get_tree_sitter_grep_args(aggregated_queries: &AggregatedQueries) -> tree_sitter_grep::Args {
+    tree_sitter_grep::Args::parse_from([
         "tree_sitter_grep",
         "-q",
         &aggregated_queries.query_text,
         "-l",
         "rust",
         "--capture",
         CAPTURE_NAME_FOR_TREE_SITTER_GREP,
-    ]);
-    let reported_any_violations = AtomicBool::new(false);
-    tree_sitter_grep::run_with_callback(
-        tree_sitter_grep_args,
-        |capture_info, file_contents, path| {
-            let (rule_index, rule_listener_index) =
-                aggregated_queries.pattern_index_lookup[capture_info.pattern_index];
-            let rule = &resolved_rules[rule_index];
-            let listener = &rule.listeners[rule_listener_index];
-            (listener.on_query_match)(
-                capture_info.node,
-                &QueryMatchContext::new(
-                    path,
-                    file_contents,
-                    rule,
-                    &reported_any_violations,
-                    &context,
-                ),
-            );
+    ])
+}
+
+fn write_files<'a>(files_to_write: impl Iterator<Item = (&'a Path, &'a [u8])>) {
+    for (path, file_contents) in files_to_write {
+        fs::write(path, file_contents).unwrap();
+    }
+}
+
+type RuleName = String;
+
+fn apply_fixes(file_contents: &mut Vec<u8>, pending_fixes: HashMap<RuleName, Vec<PendingFix>>) {
+    let non_conflicting_sorted_pending_fixes =
```

Comment:
Haven't tested this at all but per [this article](https://www.geeksforgeeks.org/check-if-any-two-intervals-overlap-among-a-given-set-of-intervals/) I took the basic approach of "sort first" for detecting "overlapping fixes"

Where per how I think ESLint behaves the basic idea is "if conflicting fixes are reported then apply a subset of them that is conflict-free" (and then presumably loop to allow the ones that didn't get applied to "get another chance" (against the updated file contents))

Specifically I interpreted that as "keep track of "groups" of fixes reported per-rule, and "apply fixes by group" (not in terms of the sequence of application, but in terms of the granularity of application)

ie detect whether the group of fixes reported by a rule _internally conflict with each other_ and if so consider that problematic/"buggy" (on the part of the rule)